express_relay/sdk/python/express_relay_utils/express_relay_types.py

```diff
@@ -0,0 +1,336 @@
+        return self
+
+
+class OpportunityBid(BaseModel):
+    """
+    Attributes:
+        opportunity_id: The ID of the opportunity.
+        amount: The amount of the bid in wei.
+        executor: The address of the executor.
+        permission_key: The permission key to bid on.
+        signature: The signature of the bid.
+        valid_until: The unix timestamp after which the bid becomes invalid.
+    """
+
+    opportunity_id: UUIDString
+    amount: IntString
+    executor: Address
+    permission_key: HexString
+    signature: SignedMessageString
+    valid_until: IntString
+
+    model_config = {
+        "arbitrary_types_allowed": True,
+    }
+
+
+class OpportunityParamsV1(BaseModel):
+    """
+    Attributes:
+        target_calldata: The calldata for the contract call.
+        chain_id: The chain ID to bid on.
+        target_contract: The contract address to call.
+        permission_key: The permission key to bid on.
+        buy_tokens: The tokens to receive in the opportunity.
+        sell_tokens: The tokens to spend in the opportunity.
+        target_call_value: The value to send with the contract call.
+        version: The version of the opportunity.
+    """
+
+    target_calldata: HexString
+    chain_id: str
+    target_contract: Address
+    permission_key: HexString
+    buy_tokens: list[TokenAmount]
+    sell_tokens: list[TokenAmount]
+    target_call_value: IntString
+    version: Literal["v1"]
+
+
+class OpportunityParams(BaseModel):
+    """
+    Attributes:
+        params: The parameters of the opportunity.
+    """
+
+    params: Union[OpportunityParamsV1] = Field(..., discriminator="version")
+
+    def to_dict(self):
+        return self.params.to_dict()
+
+
+class Opportunity(BaseModel):
+    """
+    Attributes:
+        target_calldata: The calldata for the contract call.
+        chain_id: The chain ID to bid on.
+        target_contract: The contract address to call.
+        permission_key: The permission key to bid on.
+        buy_tokens: The tokens to receive in the opportunity.
+        sell_tokens: The tokens to spend in the opportunity.
+        target_call_value: The value to send with the contract call.
+        version: The version of the opportunity.
+        creation_time: The creation time of the opportunity.
+        opportunity_id: The ID of the opportunity.
+    """
+
+    target_calldata: HexString
+    chain_id: str
+    target_contract: Address
+    permission_key: HexString
+    buy_tokens: list[TokenAmount]
+    sell_tokens: list[TokenAmount]
+    target_call_value: IntString
+    version: str
+    creation_time: IntString
+    opportunity_id: UUIDString
+
+    @model_validator(mode="after")
+    def check_version(self):
+        if self.version not in ["v1"]:
+            warnings.warn(
+                f"Cannot handle opportunity version: {self.version}. Please upgrade your client."
+            )
+            return None
```

Comment:
according to the docs, we should either return the instance or throw an exception inside these validators. So maybe this is not the best place to handle this case. I don't know what the behavior is if we return None

Reply:
yes but i think it's cleaner to handle the verification within this type. i'll add some more customized checking to throw an error but catch this while processing.